### PR TITLE
[CARBONDATA-845] Insert Select Into Same Table

### PR DESCRIPTION
**Pull request by sounakr**
Problem : Insert Select Implementation in current Spark-2.1 prohibits to read and write into same table. 

Analysis : This restriction is imposed by spark in case InsertableRelation is extended. Current insert select implementation in spark-2.1 CarbonDatasourceHadoopRelation extends InsertableRelation and override the insert method. Spark rules in case of InsertableRelation checks if source and target tables are same, then it fails and prompts  "Cannot insert overwrite into table that is also being read from."

Fix:: Spark 1.5 implementation of insert select in carbon allows reading and writing into same table, so as Hive and Parquet. Therfore as part of current fix reimplemented insert-select logic of Spark-2.1 similar to Spark-1.5. 

 a) CarbonDatasourceHadoopRelation doesn't extends InsertableRelation.
 b) New CarbonPreInsertionCasts Rule is written for spark-2.1 (For Spark-1.5 it was already present). 
 d) This new Rule is registered CarbonSessionState, 
 c) During Insert Select processing the Logical Plan wrapped by InsertIntoCarbonTable and later execute LoadTableByInsert to process insert select.

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1406/

File: integration/spark/src/main/scala/org/apache/spark/sql/hive/CarbonPreInsertionCasts.scala [hunk 1]
```diff
@@ -0,0 +1,62 @@
+/*
```
Comment (QiangCai) [inline integration/spark/src/main/scala/org/apache/spark/sql/hive/CarbonPreInsertionCasts.scala]: InsertInto function is a part of IUD feature. No need to create a new file.

**Comment by ravipesala** [comment]
Can't we have same restriction like Spark, is it really required to override this behavior?

File: integration/spark/src/main/scala/org/apache/spark/sql/hive/CarbonPreInsertionCasts.scala (hunk 1 above)
Comment (sounakr) [inline integration/spark/src/main/scala/org/apache/spark/sql/hive/CarbonPreInsertionCasts.scala]: Done, currently to keep 1.5 and 2.1 in sync, CarbonAnalysisRules.scala will keep the rules of IUD and Insert Select. Created CarbonAnalysisRules.scala in 2.1 where currenly on Insert-Select is there as IUD is still not ported to 2.1

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1419/

**Comment by sounakr** [comment]
@ravipesala Insert Select from same table is allowed both in Hive and Parquet. Spark-5746 introduced this  behavior as they are deleting the source table data first before inserting when overwrite is turned ON. But for carbon insert select Overwrite is always false and as other data sources are allowing insert select into same table it is good to have. Also the below kind of queries will help to load data. 
insert into tab1 sel a *4, b*4 from tab1;

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/DDLStrategy.scala [hunk 2]
```diff
@@ -54,6 +54,9 @@ class DDLStrategy(sparkSession: SparkSession) extends SparkStrategy {
             identifier.table.toLowerCase)) :: Nil
       case ShowLoadsCommand(databaseName, table, limit) =>
         ExecutedCommandExec(ShowLoads(databaseName, table.toLowerCase, limit, plan.output)) :: Nil
+      case InsertIntoCarbonTable(relation: CarbonDatasourceHadoopRelation,
+      _, child: LogicalPlan, _, _) =>
```
Comment (ravipesala) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/DDLStrategy.scala]: Please format it properly

**Comment by ravipesala** [comment]
LGTM

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1453/

**Comment by ravipesala** [comment]
LGTM

**Pull request merged**